src/common/mem.c

```diff
@@ -1573,4 +1573,428 @@ umem_tx_publish(struct umem_instance *umm, struct umem_rsrvd_act *rsrvd_act)
+			store_iod->io_regions[nr].sr_size = count << UMEM_CACHE_CHUNK_SZ_SHIFT;
+			sgl->sg_iovs[nr].iov_len          = sgl->sg_iovs[nr].iov_buf_len =
+			    count << UMEM_CACHE_CHUNK_SZ_SHIFT;
+			sgl->sg_iovs[nr].iov_buf = page_addr + map_offset;
+			nr++;
+
+			bmap &= ~mask;
+		} while (bmap != 0);
+
+next_bmap:
+		offset += UMEM_CACHE_CHUNK_SZ << UMEM_CHUNK_IDX_SHIFT;
+		page_addr += UMEM_CACHE_CHUNK_SZ << UMEM_CHUNK_IDX_SHIFT;
+	}
+	sgl->sg_nr_out = sgl->sg_nr = nr;
+	store_iod->io_nr            = nr;
+}
+
+/** This is O(n) but the list is tiny so let's keep it simple */
+static void
+page_insert_sorted(struct umem_store *store, struct umem_page *page, d_list_t *list)
+{
+	struct umem_page *other;
+
+	d_list_for_each_entry(other, list, pg_link) {
+		if (store->stor_ops->so_wal_id_cmp(store, page->pg_last_checkpoint,
+						   other->pg_last_checkpoint) < 0) {
+			d_list_add(&page->pg_link, &other->pg_link);
+			return;
+		}
+	}
+
+	d_list_add_tail(&page->pg_link, list);
+}
+
+int
+umem_cache_checkpoint(struct umem_store *store, umem_cache_wait_cb_t wait_cb, void *arg,
+		      uint64_t *out_id)
+{
+	struct umem_cache           *cache    = store->cache;
+	struct umem_page            *page     = NULL;
+	struct umem_checkpoint_data *chkpt_data_all;
+	struct umem_checkpoint_data *chkpt_data;
+	uint64_t                     committed_tx = 0;
+	uint64_t                     chkpt_id     = *out_id;
+	d_list_t                     free_list;
+	int                          i;
+	int                          rc;
+	int                          inflight = 0;
+
+	if (cache == NULL)
+		return 0; /* TODO: When SMD is supported outside VOS, this will be an error */
+
+	if (d_list_empty(&cache->ca_pgs_dirty))
+		return 0;
+
+	D_ASSERT(store != NULL);
+
+	D_INIT_LIST_HEAD(&free_list);
+	D_ALLOC_ARRAY(chkpt_data_all, MAX_INFLIGHT);
+	if (chkpt_data_all == NULL)
+		return -DER_NOMEM;
+
+	/** Setup the inflight IODs */
+	for (i = 0; i < MAX_INFLIGHT; i++) {
+		chkpt_data = &chkpt_data_all[i];
+		d_list_add_tail(&chkpt_data->cd_link, &free_list);
+		chkpt_data->cd_store_iod.io_regions = &chkpt_data->cd_regions[0];
+		chkpt_data->cd_sg_list.sg_iovs      = &chkpt_data->cd_iovs[0];
+	}
+
+	d_list_splice_init(&cache->ca_pgs_dirty, &cache->ca_pgs_copying);
+
+	/** First mark all pages in the new list so they won't be moved by an I/O thread.  This
+	 *  will enable us to continue the algorithm in relative isolation from I/O threads.
+	 */
+	d_list_for_each_entry(page, &cache->ca_pgs_copying, pg_link) {
+		/** Mark all pages in copying list first.  Marking them as waiting will prevent
+		 *  them from being moved to another list by an I/O operation.
+		 */
+		page->pg_waiting = 1;
+		if (store->stor_ops->so_wal_id_cmp(store, page->pg_last_inflight, chkpt_id) > 0)
+			chkpt_id = page->pg_last_inflight;
+	}
+
+	do {
+		/** first try to add up to MAX_INFLIGHT pages to the waiting queue */
+		while (inflight < MAX_INFLIGHT &&
```

Comment:
Ideally, the MAX_INFLIGHT should be based on the amount of data to be flushed but not the page count. Shouldn't we iterate the page list to generate an "umem_store_iod" for multiple pages?

Reply:
We can potentially look into improvements.   There is a tradeoff here.   Keeping pages separate is simpler and allows me to preallocate enough space to hold the iod for a page ahead of time as I know there will be a limit on possible iod entries within one 16 MB block.  We could certainly look at pushing multiple pages at a time.  But in that case, I think we would want to try to select the least recently updated pages first.